### PR TITLE
Add an opaque primitive that prevents optimisation

### PR DESCRIPTION
**Pull request by chambart**
For benchmarking or writing testsuites, it can be usefull to prevent any optimisation from happening.
This opaque primitive is translated to identity late in the compiler (cmmgen and bytegen).
This means that its intrinsic cost is very small, it does not prevent the back-end from doing its job, but it prevents any high level transformation across it.

The primitive is provided in pervasives with type `'a -> 'a`

Note: this is not intended as a replacement for `Obj.magic`. It does not make its uses any safer and would prevent the compiler from finding some misuses of `Obj.magic`.

### CODE REVIEW

**Comment by gasche** [comment]
Do we need this in the Pervasives namespace? Given the rather specialized use, maybe `Obj.opaque : 'a -> 'a` would be enough?

(I'm a bit confused by the remark on this not being a magic function. Of course, it has type `'a -> 'a` so it cannot be used for magic purposes. I understand that you mean that you chose this type-sound type because it would be bad magic anyway, but of course the type-sound type is the right choice anyway.)

File: testsuite/tests/basic-more/opaque_prim.ml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
+let f x = opaque x
+
+let () =
+  assert(f f == f);
+  assert(opaque 1 = 1);
+  assert(opaque 1. = 1.)
+
```
Comment (Drup) [inline testsuite/tests/basic-more/opaque_prim.ml]: This tests that  it's the identity, but it doesn't seem to test that it is indeed opaque, right ?

**Comment by dbuenzli** [comment]
> Do we need this in the Pervasives namespace? Given the rather specialized use, maybe Obj.opaque : 'a -> 'a would be enough?

I understand the concern about adding to `Pervasives` but if I understand this correctly it is something that acts on expressions, not on the representation of values which is the concern of the `Obj` module. 

I find the name `opaque` at the API level a bit unfortunate as there are other uses of this concept in the system and it's a very generic name. Here's a quick brain dump: `opaque_expr`, `{block,inhibit,no}_{optims,flambda}`.

**Comment by c-cube** [comment]
I think rust uses the term "blackbox" for the corresponding primitive. I would also prefer not to have it in pervasives.

File: testsuite/tests/basic-more/opaque_prim.ml (hunk 1 above)
Comment (chambart) [inline testsuite/tests/basic-more/opaque_prim.ml]: No, it's not really possible. It mainly test that this primitive is handled by the code generator since it is quite easy to add a primitive in lambda and forget it elsewhere.

Comment (Drup) [inline testsuite/tests/basic-more/opaque_prim.ml]: I guess looking at the emitted IR is far too brittle ? It's to be expected, but I feel like this could break quite easily.

**Comment by chambart** [comment]
I don't want it in `Obj`, since it's safe. And I can't see any module where it will fit.

We will try to think a bit more about the naming. We may probably start by merging it without adding it to the stdlib yet to allow merging the remaining of flambda while avoiding to commit on the API right now.

@gasche the remark about obj was to respond before someone suggest that `Obj.magic` should be `%opaque` instead of `%identity`

**Comment by gasche** [comment]
Yep, having the primitive in the compiler but no stdlib external for it would be fine. Thanks also for the `Obj.magic` clarification.

**Comment by chambart** [comment]
One possibility might be to have it in an external library like `ocplib-endian` with some kind of backward compatible version implemented as

```
let opaque x = !(ref x)
```

Which will have almost the same effect on OCaml < 4.03.

File: testsuite/tests/basic-more/opaque_prim.ml (hunk 1 above)
Comment (chambart) [inline testsuite/tests/basic-more/opaque_prim.ml]: Not really, you would have to add explicitly some code to break it. It is quite difficult to do something like that without noticing...

**Comment by lpw25** [comment]
> I find the name opaque at the API level a bit unfortunate as there are other uses of this concept in the system and it's a very generic name.

The choice of name is deliberate because it has the same meaning as those other uses. The `-opaque` option for `.ml` files is like calling the `opaque` function on the whole module.

**Comment by c-cube** [comment]
If there is a "%opaque" primitive, I think most library authors will provide their own `external` alias for it. Or even, people might use local functions, with a monomorphic type, as is the case for "%identity" (when you need to lie a bit to the typechecker but in a controlled way).

**Comment by mshinwell** [comment]
@lpw25 was, if I remember correctly, sure that this was the correct name given the semantics.  Leo, please comment.  In particular I think he said that "blackbox" was a worse name.

@dbuenzli : "opaque" is not some kind of magic construct working on expressions.  It can be thought of as a function call where the compiler knows that it is not to permit any subsequent optimization to see the result of the function application (and to believe that the computation of such result had a side effect).

This construct needs to be exposed somewhere user-facing (not in Obj), not as a hidden primitive (apart from anything else, users should not be exposed to % primitives).  We believe it is likely to be important for people writing benchmarks and test cases.  My feeling is that Pervasives is the appropriate place since it is actually a builtin like, e.g., "raise".  However if people feel strongly we could maybe consider making a new stdlib module for it.  (I would point out that whilst I see objections to putting it in Pervasives on the comments above, I see no concrete reasons for not doing so.)

**Comment by dbuenzli** [comment]
> The choice of name is deliberate because it has the same meaning as those other uses. The -opaque option for .ml files is like calling the opaque function on the whole module.

Yes, that didn't occur to me. I'm fine with `Pervasives.opaque` then.

**Comment by chambart** [comment]
@c-cube, No, users shouldn't add their own external for % primitive, they are not part of the language and are implementation details subject to change at any minor revision.

**Comment by lpw25** [comment]
> "opaque" is not some kind of magic construct working on expressions.

To make things clearer, the following will still inline (if possible) the call to `f`:

``` ocaml
opaque (f 3)
```

whereas this will not:

``` ocaml
(opaque f) 3
```

`opaque` is simply a function whose definition the optimizer cannot see.

**Comment by c-cube** [comment]
Ok, then I suppose it's fine to put it in Pervasives. After all it's less annoying than a new type (because you can always shadow values).

Side question about "%identity" (sorry for the noise): if, sometimes, I want to lie about type parameters (variance, etc.), should I use `Obj.magic` rather than "%identity"?

**Comment by dbuenzli** [comment]
@lpw25 actually that was my intuition, my mentioning of the term _expression_ was a bit unfortunate I just wanted to make a distinction that I felt that this was not `Obj` related (whose duty is to operate on the runtime representation of values).

**Comment by xavierleroy** [comment]
Do you really need "opaque" to be a %-primitive, or would a C external function (implemented as the identity function) suffice?  (Just so that no changes in the compiler back-end are needed.)

**Comment by mshinwell** [comment]
Hmm, an interesting suggestion, but I'm not sure it will be satisfactory.  For example, if you want to benchmark a pure computation within a loop, you might use "opaque" to stop the inside of the loop from being optimized away.  However, if "opaque" is an external function call, there will be an additional cost recorded together with what you're trying to measure.

The backend change is just this in Cmmgen.transl_prim_1:
    Pidentity | Popaque ->
      transl env arg

**Comment by mshinwell** [comment]
Actually, another reason for not using a C function is that it makes the compiler code more fragile.  We have exhaustive matches on the toplevel constructor of primitives (Popaque for example) in Flambda, but we can't reasonably do that for Pccall.

**Comment by gasche** [comment]
I still think that adding `opaque` to `Pervasives` is a bad idea -- veto!

One thing that I do not understand is: what is the user story for using `%opaque`? Who will use it in practice? Given the (non)-specification, my understanding is that it is useful to people testing compiler optimizations (so Pierre, Mark, and people willing to experiment with flambda and contribute improvements). Given that interpretation, having to define your own `external` to use the feature is rather ok. Now the discussion seems to be going as if arbitrary library authors were supposed to have a use for `%opaque` for their own performance measurements. Is that plausible? When would they do that? Do we have a convincing use-case for this?

Pierre, Mark, could you give examples of when you are using `%opaque` and what it is useful for?

**Comment by bobzhang** [comment]
how about supporting `[@noinline]` attribute instead?

**Comment by c-cube** [comment]
@gasche I think the user story is the following: you want to benchmark some data structure in batteries (or any other regular, user-land library), by repeatedly computing something and doing nothing with it:

``` ocaml
let tbl = MyShinyHashtbl.create ()in
for i = 0 to 100_000 do
  MyShinyHashtbl.add tbl i ();
done;
()  (* do nothing with the tbl *)
```

Since this could possibly be entirely optimized out after inlining, the only side effect being CPU temperature, you need to put `opaque` somewhere. Rust has a [similar feature](https://doc.rust-lang.org/book/benchmark-tests.html) for benchmarking (look for "test::black_box").

**Comment by Drup** [comment]
Can someone explain clearly what's the rational behind not wanting this in pervasive ?

(Other that "Wow, Pervasive, NO")

**Comment by mshinwell** [comment]
@gasche : I'm afraid just writing "veto" doesn't help us to understand what your objection to having this in Pervasives actually _is_...

There is a written specification for "opaque": it's in the pervasives.mli diff attached to this PR.  If that documentation is inadequate I am happy to receive suggestions as to how to improve it.  The use case is not something obscure for compiler writers: it's for the working programmer writing tests and benchmarks.  An example is a loop that contains a pure computation whose result is ignored for the purpose of the benchmark.  The Flambda simplifier will see that the variable bound to the result of such computation is unused, and seeing that the computation is pure (perhaps it's done some inlining in order to deduce that, but inlining is actually orthogonal to the behaviour of "opaque"), it will delete the computation itself.  By instead passing the result to "opaque" it will be made certain that the computation is executed.

**Comment by gasche** [comment]
People have plenty of reasons to define a variable named `opaque` (this is a common english names with a variety of meanings). If you add `opaque` to pervasives, everyone seeing this variable name now has to do the extra work of wondering whether this is the stdlib's one or a locally-defined one.

I certainly frown if I see someone rebinding `min` or `max` locally. Adding stuff to Pervasives does reserve some part of the namespace in practice. I don't think the use-case for `opaque` justifies doing this for everyone. This is a niche feature, it should be in a niche namespace.

The same argument could have applied to `raise_notrace`, except the risk of collision is very low for this one. If you called this `flambda_opaque` or whatever the argument would be weaker. But I still wouldn't see the reason for having this in `Pervasives`.

As a meta-point, I think this discussion is getting the "burden of proof" wrong. We should not discuss reasons for _not_ having this in Pervasives, what we should discuss is very, very good reasons to _have_ this in Pervasives. I don't see any in this thread.

> There is a written specification for "opaque": it's in the pervasives.mli diff attached to this PR. If that documentation is inadequate I am happy to receive suggestions as to how to improve it.

As a non-flambda-expert I don't see how opaque should be applied. I assume that if you write `let x = expr in opaque x`, this is not equivalent to `opaque expr`, correct? Then what about `opaque (f expr1 expr2)` where `f : int -> int -> int`. Is this equivalent to `opaque (f (opaque expr1) (opaque expr2))`, or should I write the later for extra safety? Or should I write `(opaque f) (opaque expr1) (opaque expr2)` instead?

I stand my ground: if people believe that they are expert enough to understand how to use `opaque`, they can bind an `external` value first. (Pierre's argument that `%opaque` is not part of the language and may change seems pretty thin, as I don't see how `Pervasives.opaque` would have a more robustly defined meaning.)

**Comment by c-cube** [comment]
how about this? ;)

``` ocaml
val i_swear_this_expression_is_used : 'a -> 'a
```

**Comment by dbuenzli** [comment]
> I stand my ground: if people believe that they are expert enough to understand how to use opaque, they can bind an external value first.

`opaque` can certainly be better documented, but I don't think  you need to be an "expert" to understand it. 

Writing benchmarking code fits well into the tasks of the average programmer and should not require to bind to an external, even more so since those are completely undocumented. Another use case for `opaque` is cloaking computations. 

I don't care if it's in `Pervasives` or another module but requiring externals for this doesn't sound like a user-friendly approach for the end-users of the system.

Regarding my first assertion, let's put that understandability to test: 

> As a non-flambda-expert I don't see how opaque should be applied. I assume that if you write let x = expr in opaque x, this is not equivalent to opaque expr, correct? 

I would say the answer is no, both are equivalent. 

> Then what about opaque (f expr1 expr2) where f : int -> int -> int. Is this equivalent to opaque (f (opaque expr1) (opaque expr2)), or should I write the later for extra safety? 

This is equivalent to `(f expr1 expr2)`. 

Basically if you want to write an expression where you don't want to flambda to fiddle with you write an opaque thunk: 

```
opaque @@ fun () -> expr
```

Correct ?

**Comment by gasche** [comment]
(What do you mean by "cloaking computations"?)

I doubt that an easy-to-understand semantics can be given for `opaque`, that does not depend on understanding how implementation aspects of how flambda's value analysis works. Advertising it for users that write benchmarks sounds more likely to help them shoot themselves in the foot, rather than making their life easier. 

As far as I know, most benchmarking libraries out there provide some sort of `measure : conf -> ('a -> 'b) -> 'a -> 'b` function that measure the time taken by applying the passed function to the passed argument. I hope that users of such libraries do not need to learn about `opaque`, only their developers do. Those can use an external. (Or, indeed, another module, but I don't see a good candidate besides `Obj`.)

We might have an issue with people writing bare OCaml code for performance measurement, if the way this code is written is not robust with respect to more clever compiler optimisations. Do we have an example of such code, of how it should be updated to use `opaque` or a benchmarking library? Is this code written by non-expert users¹, or is it in `operf-micro`?

¹: this is not very wise given the non-trivial amount of knowledge about reproducibility and Gc interference and statistics that can/should be packed in good benchmarking libraries, and whose absence easily leads poorly-written benchmarks to the wrong conclusion.

**Comment by dbuenzli** [comment]
> (What do you mean by "cloaking computations"?)

Masking the actual amount of work you perform. 

> Those can use an external.

Externals are not part of the language and again, are not documented. I want to work with the language as much as possible, even if I'm writing a benchmarking library.

**Comment by mshinwell** [comment]
As regards the name, I had a quick look through a part of the Jane Street tree that contains about 450,000 functions, and I only found one value called "opaque".  It wasn't a function either.  So I don't think it's really a common name.

**Comment by yallop** [comment]
> I doubt that an easy-to-understand semantics can be given for opaque

My understanding (from reading this thread but not the actual diff) is that:
- to the optimizer `opaque` behaves like an unknown function
- at runtime `opaque` disappears altogether

So calling `opaque` inhibits optimization in exactly the same way as calling a C function, since the definition of the function is invisible to the optimizer in both cases.

> `let x = expr in opaque x`, this is not equivalent to `opaque expr`, correct?

They should be equivalent.

> Then what about `opaque (f expr1 expr2)` where `f : int -> int -> int`. Is this equivalent to `opaque (f (opaque expr1) (opaque expr2))`

They're not quite equivalent: in the first example after inlining `f` the optimizer might be able to further inline (parts of) `expr1` and `expr2` in the body.  In the second case the optimizer can inline `f`, but the inner calls to `opaque` prevent any further reduction.

> Or should I write `(opaque f) (opaque expr1) (opaque expr2)` instead?

That would have the additional effect of preventing the inlining of `f` altogether.

**Comment by lpw25** [comment]
> I doubt that an easy-to-understand semantics can be given for opaque

Really, because they seem pretty clear to me. `opaque` is a function whose definition is not visible to the optimiser. Any optimisation that would require knowing that `opaque` is identity will not happen.

>  I assume that if you write let x = expr in opaque x, this is not equivalent to opaque expr, correct?

Of course they are equivalent, `opaque` is just a regular function. Again `opaque` is _not_ magic.

> Then what about opaque (f expr1 expr2) where f : int -> int -> int. Is this equivalent to opaque (f (opaque expr1) (opaque expr2))?

No, they are not equivalent. In the first case you could specialise `f` based on the known values of the arguments. For example,

``` ocaml
let f x y = y
let _ = opaque (f x y)
```

can be reduced to:

``` ocaml
let _ = opaque y
```

> Or should I write (opaque f) (opaque expr1) (opaque expr2) instead?

This completely prevents any changes by the optimiser because it is now the application of an unknown function to two unknown arguments. You could get the same with `(opaque f) expr1 expr2` since you also can't do anything with the application of an unknown function to two known arguments.

**Comment by hcarty** [comment]
@mshinwell One likely source of name collisions with opaque would be code involving colors or light.  I've used it a few times as a function to drop transparency from a RGBA-like color definition, for example.

**Comment by lpw25** [comment]
Entertainingly, the fact that @yallop's post is almost identical to mine (posted before his had appeared in my browser) demonstrates that the semantics are clearly understandable to someone with no inside knowledge of flambda.

**Comment by mshinwell** [comment]
@yallop Would you mind terribly if we used some of your wording for the comment in the .mli?

**Comment by mshinwell** [comment]
@hcarty That is an interesting point.
We could consider something like "optimizer_opaque", but we're keen on the "opaque" bit.

**Comment by yallop** [comment]
@mshinwell: you're very welcome to use it!

**Comment by gasche** [comment]
Based on @lpw25 and @yallop's replies, here would be a recommendation for the documentation. It currently says

``` ocaml
(** Semantically, [opaque v] behaves as the identity function; but when
    used with Flambda mode, for the purposes of optimisation passes such
    expression is side-effecting and yields an unknown result.  This may
    be used to prevent necessary code in applications such as benchmarks from
    being optimised away. *)
```

If the specification we want to give is about `opaque`, the function, then the documentation should not be about `opaque v`, the expression, as it would suggest that it has a non-function semantics (like `lazy v` for example). (Also `v` may or may not suggest that this is a meta-variable for a value rather than an arbitrary expression.)

(I still haven't seen any argument for why `opaque` would deserve to be in `Pervasives`. It does not.)

**Comment by hcarty** [comment]
I agree that `opaque` as part of the name makes sense here.  Since the documentation calls out flambda directly, would it make sense to add `Flambda.opaque` to the stdlib?  Or `reachable_opaque`/`keep_opaque` if the function is going to stay in `Pervasives`.

**Comment by trefis** [comment]
> I still haven't seen any argument for why `opaque` would deserve to be in `Pervasives`. It does not

What's your counter-proposition? I think I only saw `Obj` mentioned (before being shot down) in this thread.

**Comment by mshinwell** [comment]
@hcarty This shouldn't really be specific to Flambda.  In the future, we hope that Flambda will be the default (and maybe the only option in the longer term).

Pervasives seems the only logical place to put "opaque" without introducing a new stdlib module, which is surely less satisfactory.  Especially if it has some slightly different name I just cannot imagine why this is so problematic.

**Comment by gasche** [comment]
> What's your counter-proposition?

I don't have to have one. As far as I'm _personally_ concerned, `Obj.opaque` is fine, but if people don't agree on something not in Pervasives they can just bind it from an external library (either locally with an external, or using a dedicated library just as `libendian`). It does not have to be in the standard library.

(The argument about `external` not being part of the language does not mean much to me, given that `opaque` is currently specified in terms of Flambda which is also not part of the language. In any case, this is not my problem.)

**Comment by dbuenzli** [comment]
`Camlinternal.opaque` ?  `Camlinternal` doesn't exist yet but it's unlikely to clash with user code.

**Comment by avsm** [comment]
`Obj.opaque` seems entirely out of place.  The module exists for `operations on internal representations of values` and `not for the casual user`. `opaque` does neither of those, and is intended for the working programmer writing tests.  Something like `Camlinternal` or similar would seem more appropriate than dumping this in `Obj`.

**Comment by gasche** [comment]
I'm fine with `Camlinternal.opaque` -- adding modules is frowned upon but arguably the `Camlinternal*` namespace is reserved.

**Comment by dbuenzli** [comment]
> Flambda which is also not part of the language

That actually depends on the semantics you attribute to the language. In any case having it as an external only also prevents a lot of the advantages of having it as an OCaml value: it can show up in OCaml source code analysis tools, it can be deprecated etc.

**Comment by mshinwell** [comment]
As far as I know, Camlinternal is the wrong name.  The Camlinternal modules are (I think) all dependencies of code generated by the compiler in certain situations, e.g. for object support, and are not designed for general users to call directly.

**Comment by lpw25** [comment]
How about `opaque_identity`? (I wouldn't mind a corresponding `identity` as well).

**Comment by xavierleroy** [comment]
Sys.opaque ??  (Not great, I know.)

**Comment by mshinwell** [comment]
Well, actually, there are already things like Sys.max_array_length and Sys.ocaml_version, so it's not like that module is really just system/C library calls (despite what the comment at the top might imply)...

**Comment by damiendoligez** [comment]
I'd be happy with:

```
    Pervasives.opaque_identity
    Pervasives.unoptimized_identity
    Sys.opaque
    Sys.opaque_identity
    Sys.unoptimized_identity
```

I'd be OK with:

```
    Obj.opaque
    Obj.opaque_identity
    Obj.unoptimized_identity
    Pervasives.opaque
```

I'd be unhappy with `Pervasives.opaque` or a bare primitive that people have to declare as external.

@gasche, any preference among my first set?

**Comment by dbuenzli** [comment]
`Sys.opaque_identity` or `Pervasives.opaque_identity` make most sense to me (I find `unoptimized_identity`misleading).

**Comment by gasche** [comment]
I would prefer `Sys.opaque_identity` (or `Sys.opaque`; I find the longer name clearer and I like clear names, but apparently Mark wants something short) to `Pervasives.opaque_identity`, but I would not veto the latter if people have strong preferences for it.

**Comment by chambart** [comment]
I'll wait a bit until a consensus is reached and update this PR.

**Comment by mshinwell** [comment]
I didn't actually mean to imply that I wanted a short name.

**Comment by mshinwell** [comment]
@chambart I will update the comment based on @yallop 's text in due course too

**Comment by mshinwell** [comment]
Having thought about this some more, if this is going to have the longer name, I think the argument for putting it in Pervasives is stronger.  Sys isn't really very discoverable for a user.

So I am going to assert that the current proposal is for Pervasives.opaque_identity.  Any advances on this?

**Comment by c-cube** [comment]
I agree `opaque_identity` is specific enough that it won't cause weird conflicts, and the name is descriptive, so this is ok by me.

**Comment by xavierleroy** [comment]
Sys is as easily "discoverable" as List, Map and Set, and already contains some very widely used functions and constants -- certainly more used than the one we're discussing.  

I still dislike the idea of polluting Pervasives (= everybody's initial name space) with this function, no matter how long its name is.

**Comment by hcarty** [comment]
`Sys` seems like a reasonable home for this function. If other, similar interactions with compiler internals come along later do we want to add all of them to `Pervasives`?

**Comment by mshinwell** [comment]
Well, I'm not sure, but I don't think I would naturally look in Sys (I usually think of Sys as for things such as remove_file, or at least I did until I discovered the max_array_length constant etc).  It's kind of a shame there isn't another module which is "things about the compiler and runtime system".

However, now @xavierleroy has spoken, I propose we make this Sys.opaque_identity and stop this discussion.  @damiendoligez please approve.

**Comment by dbuenzli** [comment]
Le jeudi, 14 janvier 2016 à 15:51, Mark Shinwell a écrit :

> It's kind of a shame there isn't another module which is "things about the compiler and runtime system".

Given what is already in there just take Sys to be that module. I think the file system and signal stuff should be deprecated and migrated to the misnamed (as it provides a pretty good Windows compat layer it seems) unix library.

Daniel

**Comment by damiendoligez** [comment]
OK for merging with `Sys.opaque_identity`.

**Comment by mshinwell** [comment]
@chambart I will do the change to the comment, move the thing to Sys and then merge this.
